crates/bevy_math/src/curve/derivatives/mod.rs

```diff
@@ -0,0 +1,222 @@
+//! This module holds traits related to extracting derivatives from curves. In
+//! applications, the derivatives of interest are chiefly the first and second;
+//! in this module, these are provided by the traits [`CurveWithDerivative`]
+//! and [`CurveWithTwoDerivatives`].
+//!
+//! These take ownership of the curve they are used on by default, so that
+//! the resulting output may be used in more durable contexts. For example,
+//! `CurveWithDerivative<T>` is not dyn-compatible, but `Curve<WithDerivative<T>>`
+//! is, so if such a curve needs to be stored in a dynamic context, calling
+//! [`with_derivative`] and then placing the result in a
+//! `Box<Curve<WithDerivative<T>>>` is sensible.
+//!
+//! On the other hand, in more transient contexts, consuming a value merely to
+//! sample derivatives is inconvenient, and in these cases, it is recommended
+//! to use [`by_ref`] when possible to create a referential curve first, retaining
+//! liveness of the original.
+//!
+//! This module also holds the [`SampleDerivative`] and [`SampleTwoDerivatives`]
+//! traits, which can be used to easily implement `CurveWithDerivative` and its
+//! counterpart.
+//!
+//! [`with_derivative`]: CurveWithDerivative::with_derivative
+//! [`by_ref`]: Curve::by_ref
+
+pub mod adaptor_impls;
+
+use crate::{
+    common_traits::{HasTangent, WithDerivative, WithTwoDerivatives},
+    curve::{Curve, Interval},
+};
+use core::ops::Deref;
+
+#[cfg(feature = "bevy_reflect")]
+use bevy_reflect::{FromReflect, Reflect};
+
+/// Trait for curves that have a well-defined notion of derivative, allowing for
+/// derivatives to be extracted along with values.
+pub trait CurveWithDerivative<T>: SampleDerivative<T>
```

Comment:
This docstring should probably mention how to implement this trait. (Via SampleDerivative.)